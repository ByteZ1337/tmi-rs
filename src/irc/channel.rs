use crate::common::Span;

/// #channel <rest>
#[inline(always)]
pub(super) fn parse(src: &str, pos: &mut usize) -> Option<Span> {
  match src[*pos..].starts_with('#') {
    true => {
      let start = *pos;
      match src[start..].find(' ') {
        Some(end) => {
          let end = start + end;
          *pos = end + 1;
          Some(Span::from(start..end))
        }
        None => {
          let end = src.len();
          *pos = end;
          Some(Span::from(start..end))
        }
      }
    }
    false => None,
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn channel() {
    let data = "#channel <rest>";
    let mut pos = 0;

    let channel = parse(data, &mut pos).unwrap();
    assert_eq!(channel.get(data), "#channel");
    assert_eq!(&data[pos..], "<rest>");
  }
}
