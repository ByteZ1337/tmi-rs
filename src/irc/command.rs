use std::fmt::Display;

use crate::common::Span;

#[derive(Clone, Copy)]
pub(super) enum RawCommand {
  Ping,
  Pong,
  Join,
  Part,
  Privmsg,
  Whisper,
  Clearchat,
  Clearmsg,
  GlobalUserState,
  Notice,
  Reconnect,
  RoomState,
  UserNotice,
  UserState,
  Capability,
  RplWelcome,
  RplYourHost,
  RplCreated,
  RplMyInfo,
  RplNamReply,
  RplEndOfNames,
  RplMotd,
  RplMotdStart,
  RplEndOfMotd,
  Other(Span),
}

impl RawCommand {
  #[inline]
  pub(super) fn get<'src>(&self, src: &'src str) -> Command<'src> {
    match self {
      RawCommand::Ping => Command::Ping,
      RawCommand::Pong => Command::Pong,
      RawCommand::Join => Command::Join,
      RawCommand::Part => Command::Part,
      RawCommand::Privmsg => Command::Privmsg,
      RawCommand::Whisper => Command::Whisper,
      RawCommand::Clearchat => Command::ClearChat,
      RawCommand::Clearmsg => Command::ClearMsg,
      RawCommand::GlobalUserState => Command::GlobalUserState,
      RawCommand::Notice => Command::Notice,
      RawCommand::Reconnect => Command::Reconnect,
      RawCommand::RoomState => Command::RoomState,
      RawCommand::UserNotice => Command::UserNotice,
      RawCommand::UserState => Command::UserState,
      RawCommand::Capability => Command::Capability,
      RawCommand::RplWelcome => Command::RplWelcome,
      RawCommand::RplYourHost => Command::RplYourHost,
      RawCommand::RplCreated => Command::RplCreated,
      RawCommand::RplMyInfo => Command::RplMyInfo,
      RawCommand::RplNamReply => Command::RplNames,
      RawCommand::RplEndOfNames => Command::RplEndOfNames,
      RawCommand::RplMotd => Command::RplMotd,
      RawCommand::RplMotdStart => Command::RplMotdStart,
      RawCommand::RplEndOfMotd => Command::RplEndOfMotd,
      RawCommand::Other(span) => Command::Other(&src[*span]),
    }
  }
}

/// A Twitch IRC command.
#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Command<'src> {
  /// Ping the peer
  Ping,
  /// The peer's response to a [`Command::Ping`]
  Pong,
  /// Join a channel
  Join,
  /// Leave a channel
  Part,
  /// Send a message to a channel
  Privmsg,
  /// Send a private message to a user
  Whisper,
  /// Purge a user's messages in a channel
  ClearChat,
  /// Remove a single message
  ClearMsg,
  /// Sent upon successful authentication (PASS/NICK command)
  GlobalUserState,
  /// General notices from the server
  Notice,
  /// Rejoins channels after a restart
  Reconnect,
  /// Identifies the channel's chat settings
  RoomState,
  /// Announces Twitch-specific events to the channel
  UserNotice,
  /// Identifies a user's chat settings or properties
  UserState,
  /// Requesting an IRC capability
  Capability,
  // Numeric commands
  /// `001`
  RplWelcome,
  /// `002`
  RplYourHost,
  /// `003`
  RplCreated,
  /// `004`
  RplMyInfo,
  /// `353`
  RplNames,
  /// `366`
  RplEndOfNames,
  /// `372`
  RplMotd,
  /// `375`
  RplMotdStart,
  /// `376`
  RplEndOfMotd,
  /// Unknown command
  Other(&'src str),
}

impl<'src> Display for Command<'src> {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    f.write_str(self.as_str())
  }
}

impl<'src> Command<'src> {
  /// Get the string value of the [`Command`].
  pub fn as_str(&self) -> &'src str {
    use Command::*;
    match self {
      Ping => "PING",
      Pong => "PONG",
      Join => "JOIN",
      Part => "PART",
      Privmsg => "PRIVMSG",
      Whisper => "WHISPER",
      ClearChat => "CLEARCHAT",
      ClearMsg => "CLEARMSG",
      GlobalUserState => "GLOBALUSERSTATE",
      Notice => "NOTICE",
      Reconnect => "RECONNECT",
      RoomState => "ROOMSTATE",
      UserNotice => "USERNOTICE",
      UserState => "USERSTATE",
      Capability => "CAP",
      RplWelcome => "001",
      RplYourHost => "002",
      RplCreated => "003",
      RplMyInfo => "004",
      RplNames => "353",
      RplEndOfNames => "366",
      RplMotd => "372",
      RplMotdStart => "375",
      RplEndOfMotd => "376",
      Other(cmd) => cmd,
    }
  }
}

/// `COMMAND <rest>`
///
/// Returns `None` if command is unknown *and* empty
#[inline(always)]
pub(super) fn parse(src: &str, pos: &mut usize) -> Option<RawCommand> {
  let (end, next_pos) = match src[*pos..].find(' ') {
    Some(end) => {
      let end = *pos + end;
      (end, end + 1)
    }
    None => (src.len(), src.len()),
  };

  use RawCommand as C;
  let cmd = match &src[*pos..end] {
    "PING" => C::Ping,
    "PONG" => C::Pong,
    "JOIN" => C::Join,
    "PART" => C::Part,
    "PRIVMSG" => C::Privmsg,
    "WHISPER" => C::Whisper,
    "CLEARCHAT" => C::Clearchat,
    "CLEARMSG" => C::Clearmsg,
    "GLOBALUSERSTATE" => C::GlobalUserState,
    "NOTICE" => C::Notice,
    "RECONNECT" => C::Reconnect,
    "ROOMSTATE" => C::RoomState,
    "USERNOTICE" => C::UserNotice,
    "USERSTATE" => C::UserState,
    "CAP" => C::Capability,
    "001" => C::RplWelcome,
    "002" => C::RplYourHost,
    "003" => C::RplCreated,
    "004" => C::RplMyInfo,
    "353" => C::RplNamReply,
    "366" => C::RplEndOfNames,
    "372" => C::RplMotd,
    "375" => C::RplMotdStart,
    "376" => C::RplEndOfMotd,
    other if !other.is_empty() => C::Other(Span::from(*pos..end)),
    _ => return None,
  };

  *pos = next_pos;

  Some(cmd)
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn command() {
    let data = "PING <rest>";
    let mut pos = 0;

    let command = parse(data, &mut pos).unwrap();
    assert_eq!(command.get(data), Command::Ping);
    assert_eq!(&data[pos..], "<rest>");
  }
}
