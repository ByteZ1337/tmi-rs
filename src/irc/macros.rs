#[doc(hidden)]
#[macro_export]
macro_rules! __count {
  () => (0usize);
  ($x:tt $($xs:tt)*) => (1usize + $crate::__count!($($xs)*));
}

#[macro_export]
macro_rules! whitelist {
  [$($tag:ident),*] => (
    unsafe {
      $crate::irc::Whitelist::<{$crate::__count!($($tag)*)}, _>::new({
        #[allow(unused_variables)]
        #[inline]
        |src: &str, map: &mut $crate::irc::RawTags, tag: $crate::common::Span, value: $crate::common::Span| {
          match src[tag].as_bytes() {
            $($crate::irc::tags::$tag => {map.push($crate::irc::RawTagPair($crate::irc::RawTag::$tag, value));})*
            _ => {}
          };
        }
      })
    }
  )
}
