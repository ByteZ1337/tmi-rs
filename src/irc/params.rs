use crate::common::Span;

#[inline(always)]
pub(super) fn parse(src: &str, pos: &usize) -> Option<Span> {
  if !src[*pos..].is_empty() {
    Some(Span::from(*pos..src.len()))
  } else {
    None
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn params() {
    let data = ":param_a :param_b";
    let params = parse(data, &0).unwrap();
    assert_eq!(params.get(data), data)
  }
}
