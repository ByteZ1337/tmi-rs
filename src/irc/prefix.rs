use crate::common::Span;

#[derive(Debug, Clone, Copy)]
pub(super) struct RawPrefix {
  nick: Option<Span>,
  user: Option<Span>,
  host: Span,
}

impl RawPrefix {
  pub(super) fn get<'src>(&self, src: &'src str) -> Prefix<'src> {
    Prefix {
      nick: self.nick.map(|span| &src[span]),
      user: self.user.map(|span| &src[span]),
      host: &src[self.host],
    }
  }
}

// TODO: have prefix only be two variants: `User` and `Host`
/// A message prefix.
///
/// ```text,ignore
/// :nick!user@host
/// ```
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct Prefix<'src> {
  /// The `nick` part of the prefix.
  pub nick: Option<&'src str>,
  /// The `user` part of the prefix.
  pub user: Option<&'src str>,
  /// The `host` part of the prefix.
  pub host: &'src str,
}

impl<'src> std::fmt::Display for Prefix<'src> {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    match (self.nick, self.user, self.host) {
      (Some(nick), Some(user), host) => write!(f, "{nick}!{user}@{host}"),
      (Some(nick), None, host) => write!(f, "{nick}@{host}"),
      (None, None, host) => write!(f, "{host}"),
      _ => Ok(()),
    }
  }
}

/// `:nick!user@host <rest>`
#[inline(always)]
pub(super) fn parse(src: &str, pos: &mut usize) -> Option<RawPrefix> {
  if !src[*pos..].starts_with(':') {
    return None;
  }

  // :host <rest>
  // :nick@host <rest>
  // :nick!user@host <rest>
  let bytes = src.as_bytes();

  let start = *pos + 1;
  let mut host_start = start;
  let mut nick = None;
  let mut nick_end = None;
  let mut user = None;
  for i in start..bytes.len() {
    match unsafe { *bytes.get_unchecked(i) } {
      b' ' => {
        let host = Span::from(host_start..i);
        *pos = i + 1;
        return Some(RawPrefix { nick, user, host });
      }
      b'@' => {
        host_start = i + 1;
        if let Some(nick_end) = nick_end {
          user = Some(Span::from(nick_end + 1..i));
        } else {
          nick = Some(Span::from(start..i));
        }
      }
      b'!' => {
        nick = Some(Span::from(start..i));
        nick_end = Some(i);
      }
      _ => {}
    }
  }

  None
}
