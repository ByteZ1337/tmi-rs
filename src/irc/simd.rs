#[cfg(all(target_arch = "x86_64", target_feature = "sse2"))]
pub(crate) mod x86_sse;

#[cfg(all(target_arch = "x86_64", target_feature = "sse2"))]
pub(crate) use x86_sse::*;

#[cfg(all(target_arch = "x86_64", not(target_feature = "sse2")))]
const _: () = {
  compile_error!("cannot use SIMD - your CPU does not support sse2");
};

#[cfg(all(target_arch = "aarch64", target_feature = "neon"))]
pub(crate) mod arm_neon;

#[cfg(all(target_arch = "aarch64", target_feature = "neon"))]
pub(crate) use arm_neon::*;

#[cfg(all(target_arch = "aarch64", not(target_feature = "neon")))]
const _: () = {
  compile_error!("cannot use SIMD - your CPU does not support Neon");
};
