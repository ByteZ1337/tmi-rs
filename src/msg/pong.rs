//! Sent by TMI as a response to a [`Ping`][Ping].
//!
//! If the [`Ping`][Ping] contained a [`Ping::nonce`][nonce],
//! the same nonce will be set to [`Pong::nonce`].
//!
//! [Ping]: crate::msg::ping::Ping
//! [nonce]: crate::msg::ping::Ping::nonce

use super::{maybe_clone, MessageParseError};
use crate::irc::{Command, IrcMessageRef};
use std::borrow::Cow;

/// Sent by TMI as a response to a [`Ping`][Ping].
///
/// If the [`Ping`][Ping] contained a [`Ping::nonce`][nonce],
/// the same nonce will be set to [`Pong::nonce`].
///
/// [Ping]: crate::msg::ping::Ping
/// [nonce]: crate::msg::ping::Ping::nonce
#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Pong<'src> {
  #[cfg_attr(feature = "serde", serde(borrow))]
  nonce: Option<Cow<'src, str>>,
}

generate_getters! {
  <'src> for Pong<'src> as self {
    /// Unique string sent with this ping.
    nonce -> Option<&str> = self.nonce.as_deref(),
  }
}

impl<'src> Pong<'src> {
  fn parse(message: IrcMessageRef<'src>) -> Option<Self> {
    if message.command() != Command::Pong {
      return None;
    }

    Some(Pong {
      nonce: message.text().map(Cow::Borrowed),
    })
  }

  /// Clone data to give the value a `'static` lifetime.
  pub fn into_owned(self) -> Pong<'static> {
    Pong {
      nonce: self.nonce.map(maybe_clone),
    }
  }
}

impl<'src> super::FromIrc<'src> for Pong<'src> {
  #[inline]
  fn from_irc(message: IrcMessageRef<'src>) -> Result<Self, MessageParseError> {
    Self::parse(message).ok_or(MessageParseError)
  }
}

impl<'src> From<Pong<'src>> for super::Message<'src> {
  fn from(msg: Pong<'src>) -> Self {
    super::Message::Pong(msg)
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn parse_ping() {
    assert_irc_snapshot!(Pong, ":tmi.twitch.tv PONG");
  }

  #[test]
  fn parse_ping_nonce() {
    assert_irc_snapshot!(Pong, ":tmi.twitch.tv PONG :nonce");
  }

  #[cfg(feature = "serde")]
  #[test]
  fn roundtrip_ping() {
    assert_irc_roundtrip!(Pong, ":tmi.twitch.tv PONG");
  }

  #[cfg(feature = "serde")]
  #[test]
  fn roundtrip_ping_nonce() {
    assert_irc_roundtrip!(Pong, ":tmi.twitch.tv PONG :nonce");
  }
}
