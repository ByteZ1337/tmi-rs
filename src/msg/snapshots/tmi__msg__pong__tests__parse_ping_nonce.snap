---
source: src/msg/pong.rs
expression: "f(\":tmi.twitch.tv PONG :nonce\")"
---
Pong {
    nonce: Some(
        "nonce",
    ),
}
